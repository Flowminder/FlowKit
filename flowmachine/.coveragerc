[run]
branch = True

[report]
fail_under = 90

omit =
    */python?.?/*
    */venv/*
    */site-packages/nose/*
    */tests/*
    */_version.py
    */versioneer.py
    *__init__*
    flowmachine/core/server/server.py

exclude_lines =
    if __name__ == '__main__':

show_missing = True
