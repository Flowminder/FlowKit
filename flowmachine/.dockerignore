build/
dist/
tests/
Dockerfile/
.ipynb_checkpoints/
.pytest_cache/
.flowmachine.egg-info/
docs/