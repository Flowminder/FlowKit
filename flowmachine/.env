# Set matplotlib backend to prevent ImportError when running in a virtual env
MPLBACKEND="agg"