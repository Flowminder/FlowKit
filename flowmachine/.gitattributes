flowmachine/_version.py export-subst
