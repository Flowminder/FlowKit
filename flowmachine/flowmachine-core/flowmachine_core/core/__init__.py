# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

"""
Definition of the main Abstract Base Class of flowmachine_core the Query class
and associated code.
"""

from .connection import Connection
from .init import connect
from .logging import init_logging, set_log_level

sub_modules = ["errors", "mixins", "api"]

methods = [
    "Connection",
    "connect",
]

__all__ = methods + sub_modules
