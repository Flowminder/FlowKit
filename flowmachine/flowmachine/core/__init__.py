# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

"""
Definition of the main Abstract Base Class of flowmachine the Query class
and associated code.
"""

from .connection import Connection
from .query import Query
from .table import Table
from .geotable import GeoTable
from .init import connect
from .logging import init_logging, set_log_level
from .spatial_unit import make_spatial_unit
from .join_to_location import JoinToLocation, location_joined_query
from .custom_query import CustomQuery
from .grid import Grid

sub_modules = ["errors", "mixins", "api"]

methods = [
    "Query",
    "Table",
    "GeoTable",
    "Connection",
    "connect",
    "make_spatial_unit",
    "JoinToLocation",
    "location_joined_query",
    "CustomQuery",
    "Grid",
]

__all__ = methods + sub_modules
