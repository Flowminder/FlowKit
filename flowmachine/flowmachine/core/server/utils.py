# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

import zmq
import zmq.asyncio

__all__ = [
    "send_zmq_message_and_receive_reply",
    "send_zmq_message_and_await_reply",
    "FM_EXAMPLE_MESSAGE",
]


async def send_zmq_message_and_await_reply(msg, port=5555, host="localhost"):
    """
    Helper function to send JSON messages to the flowmachine server (via zmq) and await a reply.

    This is mainly useful for interactive testing and debugging asyncio behaviour.

    Parameters
    ----------
    msg : dict
        Dictionary representing a valid zmq message.

    port : int or str
        Port on which the flowmachine server is running (default: 5555)

    host : str
        The host on which the flowmachine server is running (default: 'localhost')


    Example
    -------

    >>> msg = {
    ...     "action": "run_query",
    ...     "request_id": "DUMMY_ID",
    ...      "params": {"query_kind": "daily_location", "date": "2016-01-01", "method": "last", "aggregation_unit": "admin3", "subscriber_subset": None}
    ... }

    >>> await send_zmq_message_and_await_reply(msg)
    {'status': 'accepted', 'msg': '', 'data': {'query_id': 'e39b0d45bc6b46b7700c67cd52f00455'}}

    >>> await send_zmq_message_and_await_reply({"action": "get_sql_for_query_result", "request_id": "DUMMY_ID", "params": {"query_id": "e39b0d45bc6b46b7700c67cd52f00455"}})
    {'status': 'done', 'msg': '', 'data': {'query_id': 'e39b0d45bc6b46b7700c67cd52f00455', 'sql': 'SELECT * FROM cache.xe39b0d45bc6b46b7700c67cd52f00455'}}

    """
    context = zmq.asyncio.Context.instance()
    socket = context.socket(zmq.REQ)
    socket.connect(f"tcp://{host}:{port}")
    print(f"Sending message: {msg}")
    await socket.send_json(msg)
    reply = await socket.recv_json()
    context.destroy()
    return reply


def send_zmq_message_and_receive_reply(msg, port=5555, host="localhost"):
    """
    Helper function to send JSON messages to the flowmachine server (via zmq) and receive a reply.

    This is mainly useful for interactive testing and debugging.

    Parameters
    ----------
    msg : dict
        Dictionary representing a valid zmq message.

    port : int or str
        Port on which the flowmachine server is running (default: 5555)

    host : str
        The host on which the flowmachine server is running (default: 'localhost')


    Example
    -------

    >>> msg = {
    ...     "action": "run_query",
    ...     "request_id": "DUMMY_ID",
    ...      "params": {"query_kind": "daily_location", "date": "2016-01-01", "method": "last", "aggregation_unit": "admin3", "subscriber_subset": None}
    ... }

    >>> send_zmq_message_and_receive_reply(msg)
    {'status': 'accepted', 'msg': '', 'data': {'query_id': 'e39b0d45bc6b46b7700c67cd52f00455'}}

    >>> send_zmq_message_and_receive_reply({"action": "get_sql_for_query_result", "request_id": "DUMMY_ID", "params": {"query_id": "e39b0d45bc6b46b7700c67cd52f00455"}})
    {'status': 'done', 'msg': '', 'data': {'query_id': 'e39b0d45bc6b46b7700c67cd52f00455', 'sql': 'SELECT * FROM cache.xe39b0d45bc6b46b7700c67cd52f00455'}}

    """
    context = zmq.Context.instance()
    socket = context.socket(zmq.REQ)
    socket.connect(f"tcp://{host}:{port}")
    print(f"Sending message: {msg}")
    socket.send_json(msg)
    reply = socket.recv_json()
    context.destroy()
    return reply


FM_EXAMPLE_MESSAGE = {
    "action": "run_query",
    "params": {
        "query_kind": "spatial_aggregate",
        "locations": {
            "query_kind": "daily_location",
            "date": "2016-01-01",
            "method": "last",
            "aggregation_unit": "admin3",
            "subscriber_subset": None,
        },
    },
    "request_id": "DUMMY_ID",
}
