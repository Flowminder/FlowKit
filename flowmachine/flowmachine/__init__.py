# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

# -*- coding: utf-8 -*-
"""
FlowMachine is a Python based SQL wrapper built specifically for Flowminder's
PostgreSQL CDR database: Flowdb.

It provides a standardised way of interfacing with our postgres database. 
FlowMachine provides SQL based queries that compute commonly used features and
metrics to do with CDR data, such as `DailyLocations`, `ModalLocations`,
`Flows` and `RadiusOfGyration`.

The heart of flowmachine is the AbstractBaseClass Query, which defines how an
SQL-based feature behaves. All metrics define an SQL querty, 
and inherit from FlowMachine's main `Query()` class.

"""

from .versions import __version__
from .core.init import connect, connections
from .features.utilities import GroupValues, feature_collection
from flowmachine.core.logging import init_logging
import flowmachine.models
import flowmachine.features
import flowmachine.utils
import flowmachine.core

methods = ["GroupValues", "feature_collection", "connect", "connections"]
sub_modules = ["core", "features", "utils", "models"]
__all__ = methods + sub_modules

# Initialise loggers when flowmachine is imported
init_logging()

from ._version import get_versions

__version__ = get_versions()["version"]
del get_versions
