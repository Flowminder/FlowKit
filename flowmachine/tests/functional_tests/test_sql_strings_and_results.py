# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

from flowmachine.utils import pretty_sql

from flowmachine.core import CustomQuery, make_spatial_unit
from flowmachine.features import daily_location


def test_daily_location_1_sql(diff_reporter):
    """
    Simple daily location query returns the expected SQL string.
    """
    dl = daily_location("2016-01-01", "2016-01-02")
    sql = pretty_sql(dl.get_query())
    diff_reporter(sql)


def test_daily_location_1_df(get_dataframe, diff_reporter):
    """
    Simple daily location query returns the expected data.
    """
    dl = daily_location("2016-01-01", "2016-01-02")
    df = get_dataframe(dl)
    diff_reporter(df.to_csv())


def test_daily_location_2_sql(diff_reporter):
    """
    Daily location query with non-default parameters returns the expected SQL string.
    """
    dl = daily_location(
        "2016-01-04",
        spatial_unit=make_spatial_unit(
            "admin", level=2, region_id_column_name="admin2pcod"
        ),
        hours=(3, 9),
        method="most-common",
        subscriber_identifier="imei",
        ignore_nulls=False,
        subscriber_subset=[
            "2GJxeNazvlgZbqj6",
            "7qKmzkeMbmk5nOa0",
            "8dpPLR15XwR7jQyN",
            "1NqnrAB9bRd597x2",
        ],
    )
    sql = pretty_sql(dl.get_query())
    diff_reporter(sql)


def test_daily_location_2_df(get_dataframe, diff_reporter):
    """
    Daily location query with non-default parameters returns the expected data.
    """
    dl = daily_location(
        "2016-01-04",
        spatial_unit=make_spatial_unit("admin", level=2),
        hours=(3, 9),
        method="most-common",
        # subscriber_identifier="imei",
        ignore_nulls=False,
        subscriber_subset=[
            "2GJxeNazvlgZbqj6",
            "7qKmzkeMbmk5nOa0",
            "8dpPLR15XwR7jQyN",
            "1NqnrAB9bRd597x2",
        ],
    )
    df = get_dataframe(dl)
    diff_reporter(df.to_csv())


def test_daily_location_3_sql(diff_reporter):
    """
    Daily location query with non-default parameters returns the expected data.
    """
    subset_query = CustomQuery(
        "SELECT DISTINCT msisdn AS subscriber FROM events.calls WHERE msisdn in ('GNLM7eW5J5wmlwRa', 'e6BxY8mAP38GyAQz', '1vGR8kp342yxEpwY')",
        column_names=["subscriber"],
    )
    dl = daily_location(
        "2016-01-05",
        spatial_unit=make_spatial_unit("cell"),
        hours=(23, 5),
        method="last",
        # subscriber_identifier="imei",
        # ignore_nulls=False,
        subscriber_subset=subset_query,
    )
    sql = pretty_sql(dl.get_query())
    diff_reporter(sql)


def test_daily_location_3_df(get_dataframe, diff_reporter):
    """
    Daily location query with non-default parameters returns the expected data.
    """
    subset_query = CustomQuery(
        "SELECT DISTINCT msisdn AS subscriber FROM events.calls WHERE msisdn in ('GNLM7eW5J5wmlwRa', 'e6BxY8mAP38GyAQz', '1vGR8kp342yxEpwY')",
        column_names=["subscriber"],
    )
    dl = daily_location(
        "2016-01-05",
        spatial_unit=make_spatial_unit("cell"),
        hours=(23, 5),
        method="last",
        # subscriber_identifier="imei",
        # ignore_nulls=False,
        subscriber_subset=subset_query,
    )
    df = get_dataframe(dl)
    diff_reporter(df.to_csv())


def test_daily_location_4_sql(diff_reporter):
    """
    Regression test; this verifies the SQL statement for the test below (which checks the resulting dataframe)
    """
    subset_query = CustomQuery(
        "SELECT * FROM (VALUES ('dr9xNYK006wykgXj')) as tmp (subscriber)",
        column_names=["subscriber"],
    )
    dl = daily_location(
        "2016-01-05",
        table="events.calls",
        hours=(22, 6),
        subscriber_subset=subset_query,
    )
    sql = pretty_sql(dl.get_query())
    diff_reporter(sql)


def test_daily_location_4_df(get_dataframe, diff_reporter):
    """
    Regression test; the expected result is empty because the given subscriber does not make any calls on the given date.
    """
    subset_query = CustomQuery(
        "SELECT * FROM (VALUES ('dr9xNYK006wykgXj')) as tmp (subscriber)",
        ["subscriber"],
    )
    dl = daily_location(
        "2016-01-05",
        table="events.calls",
        hours=(22, 6),
        subscriber_subset=subset_query,
    )
    df = get_dataframe(dl)
    diff_reporter(df.to_csv())


def test_daily_location_5_sql(diff_reporter):
    """
    Daily location query with non-default parameters returns the expected data.
    """
    subset_query = CustomQuery(
        "SELECT DISTINCT msisdn AS subscriber FROM events.calls WHERE msisdn in ('GNLM7eW5J5wmlwRa', 'e6BxY8mAP38GyAQz', '1vGR8kp342yxEpwY')",
        ["subscriber"],
    )
    dl = daily_location(
        "2016-01-05",
        spatial_unit=make_spatial_unit("cell"),
        hours=(23, 5),
        method="last",
        # subscriber_identifier="imei",
        # ignore_nulls=False,
        subscriber_subset=subset_query,
    )
    sql = pretty_sql(dl.get_query())
    diff_reporter(sql)


def test_daily_location_5_df(get_dataframe, diff_reporter):
    """
    Daily location query with non-default parameters returns the expected data.
    """
    subset_query = CustomQuery(
        """
        SELECT DISTINCT msisdn AS subscriber
        FROM events.calls
        WHERE (   (datetime >= '2016-01-01 08:00:00' AND datetime <= '2016-01-01 20:00:00' AND substring(tac::TEXT, 0, 2) = '68')
               OR (datetime >= '2016-01-07 14:00:00' AND datetime <= '2016-01-07 15:00:00' AND duration < 400))
        """,
        ["subscriber"],
    )

    dl = daily_location(
        "2016-01-02",
        spatial_unit=make_spatial_unit("admin", level=3),
        hours=(4, 9),
        method="most-common",
        # subscriber_identifier="imei",
        # ignore_nulls=False,
        subscriber_subset=subset_query,
    )
    df = get_dataframe(dl)
    diff_reporter(df.to_csv())


def test_daily_location_6_sql(diff_reporter):
    """
    Regression test; this verifies the SQL statement for the test below (which checks the resulting dataframe)
    """
    subset_query = CustomQuery(
        """
        SELECT outgoing, datetime, duration, msisdn AS subscriber
        FROM events.calls
        WHERE datetime::date = '2016-01-01' AND duration > 2000
        """,
        ["subscriber"],
    )
    dl = daily_location("2016-01-03", table="calls", subscriber_subset=subset_query)
    sql = pretty_sql(dl.get_query())
    diff_reporter(sql)


def test_daily_location_6_df(get_dataframe, diff_reporter):
    """
    Regression test; the expected result is empty because the given subscriber does not make any calls on the given date.
    """
    subset_query = CustomQuery(
        """
        SELECT outgoing, datetime, duration, msisdn AS subscriber
        FROM events.calls
        WHERE datetime::date = '2016-01-01' AND duration > 2000
        """,
        ["outgoing", "datetime", "duration", "subscriber"],
    )
    dl = daily_location("2016-01-03", table="calls", subscriber_subset=subset_query)
    df = get_dataframe(dl)
    diff_reporter(df.to_csv())


def test_versioned_site_sql(diff_reporter):
    """
    Verify the SQL for a versioned-site spatial unit.
    """
    su = make_spatial_unit("versioned-site")
    sql = pretty_sql(su.get_query())
    diff_reporter(sql)
