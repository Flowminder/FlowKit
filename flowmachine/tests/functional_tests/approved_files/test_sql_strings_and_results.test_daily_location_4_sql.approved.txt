SELECT final_time.subscriber,
       pcod
FROM (SELECT subscriber_locs.subscriber,
             time,
             pcod,
             row_number() OVER (PARTITION BY subscriber_locs.subscriber
                                ORDER BY time DESC) AS rank
      FROM (SELECT subscriber,
                   datetime AS time,
                   pcod
            FROM (SELECT l.datetime,
                         l.location_id,
                         l.subscriber,
                         sites.pcod
                  FROM (SELECT tbl.datetime,
                               tbl.location_id,
                               tbl.subscriber
                        FROM (SELECT events.calls.datetime AS datetime,
                                     events.calls.location_id AS location_id,
                                     events.calls.msisdn AS subscriber
                              FROM events.calls
                              WHERE (events.calls.datetime >= '2016-01-05 00:00:00')
                                AND (events.calls.datetime < '2016-01-06 00:00:00')
                                AND ((   (to_char(events.calls.datetime, 'HH24:MI') < '06:00')
                                      OR (to_char(events.calls.datetime, 'HH24:MI') >= '22:00')))) AS tbl
                             INNER JOIN (SELECT *
                                         FROM ((VALUES ('dr9xNYK006wykgXj'))) AS tmp(subscriber)) AS subset_query ON tbl.subscriber = subset_query.subscriber) AS l
                       INNER JOIN (SELECT loc_table.id AS location_id,
                                          loc_table.date_of_first_service,
                                          loc_table.date_of_last_service,
                                          geom_table.admin3pcod AS pcod
                                   FROM infrastructure.cells AS loc_table
                                        INNER JOIN (SELECT gid,
                                                           admin0name,
                                                           admin0pcod,
                                                           admin1name,
                                                           admin1pcod,
                                                           admin2name,
                                                           admin2pcod,
                                                           admin3name,
                                                           admin3pcod,
                                                           admin3refn,
                                                           admin3altn,
                                                           admin3al_1,
                                                           date,
                                                           validon,
                                                           validto,
                                                           shape_star,
                                                           shape_stle,
                                                           shape_leng,
                                                           shape_area,
                                                           geom
                                                    FROM geography.admin3) AS geom_table ON st_within(CAST(loc_table.geom_point AS geometry),
                                                                                                      CAST(st_setsrid(geom_table.geom, 4326) AS geometry))) AS sites ON (l.location_id = sites.location_id)
                                                                                                                                                                    AND CAST((l.datetime) AS date) BETWEEN COALESCE(sites.date_of_first_service,
                                                                                                                                                                                                                    CAST(('-infinity') AS timestamptz))
                                                                                                                                                                                                       AND COALESCE(sites.date_of_last_service,
                                                                                                                                                                                                                    CAST(('infinity') AS timestamptz))) AS foo
            WHERE (location_id IS NOT NULL)
              AND (location_id <> '')) AS subscriber_locs) AS final_time
WHERE rank = 1
